### PR TITLE
[CARBONDATA-3984] Fix compaction and longStringColumn validation issue.

### PR DESCRIPTION
**Pull request by Karan980**
### Why is this PR needed?
Fix multiple issues occurred after altering the column dataType to long_string_column.
a) When dataType of a String column which is also provided as range column in table properties is altered to long_string_column. It throws following error while performing compaction on the table.
**VARCHAR not supported for the filter expression;** 
b) When longStringColumn name is present in UpperCase then Validation check for SortColumns fails. As names of SortColumns are present in LowerCase.

 ### What changes were proposed in this PR?
a) Added condition for VARCHAR datatype in all conditional expressions.
b) Changed the long_string_columns name to lowerCase before doing validation.
   
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4055/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2317/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4067/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2328/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4068/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2329/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4076/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2336/

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala [hunk 1]
```diff
@@ -110,6 +110,21 @@ class VarcharDataTypesBasicTestCase extends QueryTest with BeforeAndAfterEach wi
     assert(exceptionCaught.getMessage.contains("its data type is not string"))
   }
 
+  test("cannot alter sort_columns dataType to long_string_columns") {
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:113]: please add test case for compaction failure case

Comment (Karan980) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:117]: Done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 2]
```diff
@@ -3344,6 +3344,9 @@ public int compare(org.apache.carbondata.format.ColumnSchema o1,
     // for setting long string columns
     if (!longStringColumnsString.isEmpty()) {
       String[] inputColumns = longStringColumnsString.split(",");
+      for (int i = 0; i < inputColumns.length; i++) {
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java:3347]: instead of forloop, you can use `Arrays.stream(inputColumns).map(longColumn => longColumn.trim().toLowerCase()).toArray(String[]::new)` in a more functional manner

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2346/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4086/

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala [hunk 3]
```diff
@@ -110,6 +114,51 @@ class VarcharDataTypesBasicTestCase extends QueryTest with BeforeAndAfterEach wi
...
+    sql("insert into long_string_table select 1, 'ab', 'cool1'")
+    sql("insert into long_string_table select 2, 'abc', 'cool2'")
+    sql("ALTER TABLE long_string_table compact 'minor'")
+
+    val carbonTable = CarbonMetadata.getInstance().getCarbonTable(
+      CarbonCommonConstants.DATABASE_DEFAULT_NAME,
+      "long_string_table"
+    )
+    val absoluteTableIdentifier = carbonTable
+      .getAbsoluteTableIdentifier
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:153]: move this line above

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala [hunk 4]
```diff
@@ -110,6 +114,51 @@ class VarcharDataTypesBasicTestCase extends QueryTest with BeforeAndAfterEach wi
...
+           | TBLPROPERTIES('SORT_COLUMNS'='name, address')
+           |""".
+          stripMargin)
+      sql("ALTER TABLE long_string_table SET TBLPROPERTIES('long_String_columns'='NAME')")
+    }
+    assert(exceptionCaught.getMessage.contains("LONG_STRING_COLUMNS cannot be present in sort columns: name"))
+  }
+
+  test("check compaction after altering range column dataType to longStringColumn") {
+    CarbonProperties.getInstance()
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:133]: this property value should be reset after this test case else it will impact later test cases. Take the exiting value, set new value, reset to old value

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala [hunk 5]
```diff
@@ -29,7 +31,9 @@ import org.apache.carbondata.common.exceptions.sql.MalformedCarbonCommandExcepti
 import org.apache.carbondata.core.constants.CarbonCommonConstants
 import org.apache.carbondata.core.metadata.CarbonMetadata
 import org.apache.carbondata.core.metadata.datatype.DataTypes
+import org.apache.carbondata.core.statusmanager.SegmentStatusManager
 import org.apache.carbondata.core.util.CarbonProperties
+
```
Comment (akashrn5) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:36]: avoid unnecessary changes

**Comment by akashrn5** [comment]
@Karan980 the PR title shouldn't be so long, give a small but descriptive title and details you can explain in PR description.

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 2 above)
Comment (Karan980) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala (hunk 3 above)
Comment (Karan980) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala]: Done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala (hunk 4 above)
Comment (Karan980) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/longstring/VarcharDataTypesBasicTestCase.scala:134]: Done

**Comment by Karan980** [comment]
> @Karan980 the PR title shouldn't be so long, give a small but descriptive title and details you can explain in PR description.

Done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4092/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2351/

**Comment by akashrn5** [comment]
LGTM
